Public/Src/Cache/ContentStore/Distributed/NuCache/ContentLocationDatabase.cs

```diff
@@ -285,25 +272,27 @@ public void GarbageCollect(OperationContext context)
 
             using (var cancellableContext = TrackShutdown(context))
             {
-                DoGarbageCollect(cancellableContext);
+                var tasks = new List<Task> { Task.Run(() => DoGarbageCollectMetadata(cancellableContext)) };
+                if (_isContentGarbageCollectionEnabled)

```

Comment:
Do you want to have _isMetadataGarbageCollectionEnabled here as well to have the full logic for the entire GC?

Reply:
Another question: do you think it make sense to separate the configuration? Maybe we can do gc for content and metadata if gc is on, without separating them?

This will make the configuration simpler but maybe less flexible. But I can't imagine (for now) the case when we want to GC one thing but not another...

---
In reply to: [309381691](https://github.com/microsoft/BuildXL/pull/679#discussion_r309381691) [](ancestors = 309381691)

Reply:
Two things here:

I've missed that metadata is collected all the time, but I'm not sure that doing the GC for both metadata and content in parallel is a good idea.

Is it actually better to do this one at a time maybe?

---
In reply to: [309382147](https://github.com/microsoft/BuildXL/pull/679#discussion_r309382147) [](ancestors = 309382147,309381691)

Reply:
Like you said, as it stands, the timer will always be on, and at most what could happen is that it triggers and does nothing; in fact, this is what will indeed happen when this code is deployed, on all nodes in the system, as the metadata column family should be empty)

Since the SQLiteMemoizationStore runs GC independently of its ContentStore, this is what this logic is mimicking. Indeed, it would be better to have 2 separate timers that can be on or off independently, but it also seems overkill for something that ought to be replaced ASAP.

It doesn't change anything to do this serially right now, because both column families are treated separately, so no -new- inconsistencies can arise. Also, right now the ContentLocationDatabase that stores the locations is different than the one that stores metadata


---
In reply to: [309384685](https://github.com/microsoft/BuildXL/pull/679#discussion_r309384685) [](ancestors = 309384685,309382147,309381691)